src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,231 @@
+                                                if (AuthenticationHelper.IsServerNonceStale(_digestResponse) &&
+                                                    await AuthenticationHelper.TrySetDigestAuthToken(request, currentCredential, _digestResponse, HttpKnownHeaderNames.Authorization).ConfigureAwait(false))
+                                                {
+                                                    response.Dispose();
+                                                    response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                                                }
+
+                                                break;
+                                            }
+                                        }
+                                    }
+                                }
+                                break;
+
+                            case AuthenticationHelper.Basic:
+                                if (AuthenticationHelper.TrySetBasicAuthToken(request, currentCredential))
+                                {
+                                    response.Dispose();
+                                    response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                                }
+                                break;
+                        }
+                    }
+                }
+
+                if (!RequestNeedsRedirect(response))
+                {
+                    break;
+                }
+
+                // Clear the authorization header, if the request requires redirect.
+                request.Headers.Authorization = null;
+
+                Uri location = response.Headers.Location;
+                if (location == null)
+                {
+                    // No location header. Nothing to redirect to.
+                    break;
+                }
+
+                if (!location.IsAbsoluteUri)
+                {
+                    location = new Uri(request.RequestUri, location);
+                }
+
+                // Disallow automatic redirection from secure to non-secure schemes
+                bool allowed =
+                    (HttpUtilities.IsSupportedNonSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedScheme(location.Scheme)) ||
+                    (HttpUtilities.IsSupportedSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedSecureScheme(location.Scheme));
+                if (!allowed)
+                {
+                    break;
+                }
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
+
+                // Set up for the automatic redirect
+                request.RequestUri = location;
```

Comment:
Does the Host header need to be cleared/reset?

Reply:
WinHttpHandler is not clearing it 

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs#L182

Reply:
> WinHttpHandler is not clearing it

Does that mean it's correct or that WinHttpHandler has a bug, too?
cc: @davidsh

Reply:
>Does that mean it's correct or that WinHttpHandler has a bug, too?

WinHttpHandler relies on native WinHTTP to manage the "Host" header.  WinHTTP "owns" that header for the most part.

Reply:
Looking at the implementation in HttpConnection, the actual host header that is sent on the wire, is value taken from uri.IdnHost. And the Host header on the request is used only for creating a ssl connection in CreateConnection. This Host header is set by the user, if this header is null then host uri.IdnHost is used. During redirect, I don't think a new connection is created? From code looks like new connection is created only for proxy. So on redirect, the host value that is sent on wire will always be the requestUri.IdnHost. The Host header is used only for opening a connection. So clearing host header is not required.

Reply:
Let's take that to a separate issue. it sounds like there may be several things wrong with the host header usage.

Reply:
> Looking at the implementation in HttpConnection, the actual host header that is sent on the wire for a request, is value taken from uri.IdnHost here.

Only if no host header was set:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs#L259-L262

Reply:
@Tratcher Are you opening an issue with details on the host header usage?

Reply:
Go ahead